### PR TITLE
Login activity with Google, Facebook, and Email-Password sign-in activities

### PR DESCRIPTION
**Pull request by jschadaway**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -18,18 +19,39 @@ android {
...
+    compile 'com.google.firebase:firebase-auth:11.6.0'
+
+    compile 'com.google.android.gms:play-services-auth:11.6.0'
+
+    compile 'com.android.support:design:26.1.0'
+    compile 'com.firebase:firebase-client-android:2.5.2+'
+    compile 'com.facebook.android:facebook-android-sdk:[4,5)'
 }
 
+
```
Comment (alexrosenfeld10) [inline app/build.gradle]: remove blank lines

Comment (jschadaway) [reply app/build.gradle]: DAMNIT

Comment (jschadaway) [reply app/build.gradle]: ecks dee

File: app/src/main/java/com/teamcaffeine/hotswap/BaseActivity.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Override
+    public void onStop() {
+        super.onStop();
+        hideProgressDialog();
+    }
+
+}
+
+///**
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/BaseActivity.java]: Can we try to not include these sort of comments?

File: app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.auth.FirebaseUser;
+
+public class EmailPasswordActivity extends BaseActivity implements
+        View.OnClickListener {
+
+    private static final String TAG = "EmailPassword";
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java]: Might want a package level Tags.java file that enumerates all tags we'll use, just to keep things clean and organized

File: app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                            // Sign in success, update UI with the signed-in user's information
+                            Log.d(TAG, "createUserWithEmail:success");
+                            FirebaseUser user = mAuth.getCurrentUser();
+                            updateUI(user);
+                        } else {
+                            // If sign in fails, display a message to the user.
+                            Log.w(TAG, "createUserWithEmail:failure", task.getException());
+                            Toast.makeText(EmailPasswordActivity.this, "Authentication failed.",
+                                    Toast.LENGTH_SHORT).show();
+                            updateUI(null);
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java]: Let's make the Toast for failed login more descriptive than "Authentication failed". Something like this, maybe:

switch(task.getException().getType() {
    case ....:
       // Set toast message here

File: app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                        if (task.isSuccessful()) {
+                            // Sign in success, update UI with the signed-in user's information
+                            Log.d(TAG, "signInWithEmail:success");
+                            FirebaseUser user = mAuth.getCurrentUser();
+                            updateUI(user);
+                        } else {
+                            // If sign in fails, display a message to the user.
+                            Log.w(TAG, "signInWithEmail:failure", task.getException());
+                            Toast.makeText(EmailPasswordActivity.this, "Authentication failed.",
+                                    Toast.LENGTH_SHORT).show();
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java]: Same idea here, as above

File: app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+                        hideProgressDialog();
+                        // [END_EXCLUDE]
+                    }
+                });
+        // [END sign_in_with_email]
+    }
+
+    private void signOut() {
+        mAuth.signOut();
+        updateUI(null);
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java]: Populate UI with a "sign out successful message"

File: app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+                        if (task.isSuccessful()) {
+                            Toast.makeText(EmailPasswordActivity.this,
+                                    "Verification email sent to " + user.getEmail(),
+                                    Toast.LENGTH_SHORT).show();
+                        } else {
+                            Log.e(TAG, "sendEmailVerification", task.getException());
+                            Toast.makeText(EmailPasswordActivity.this,
+                                    "Failed to send verification email.",
+                                    Toast.LENGTH_SHORT).show();
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/EmailPasswordActivity.java]: ^^

File: app/src/main/java/com/teamcaffeine/hotswap/FacebookLoginActivity.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                        if (task.isSuccessful()) {
+                            // Sign in success, update UI with the signed-in user's information
+                            Log.d(TAG, "signInWithCredential:success");
+                            FirebaseUser user = mAuth.getCurrentUser();
+                            updateUI(user);
+                        } else {
+                            // If sign in fails, display a message to the user.
+                            Log.w(TAG, "signInWithCredential:failure", task.getException());
+                            Toast.makeText(FacebookLoginActivity.this, "Authentication failed.",
+                                    Toast.LENGTH_SHORT).show();
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/FacebookLoginActivity.java]: Same ^^

File: app/src/main/java/com/teamcaffeine/hotswap/HomeActivity.java [hunk 9]
```diff
@@ -1,13 +1,29 @@
...
         setContentView(R.layout.activity_home);
+
+        Button logout = (Button) findViewById(R.id.btnFacebookLogout);
+        logout.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                LoginManager.getInstance().logOut();
+                Intent logout = new Intent(HomeActivity.this, LoginActivity.class);
+                startActivity(logout);
+                finish();
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/HomeActivity.java]: Nice, removes from backstack. Do we need some extra flags on the intent for this as well? I think we might

File: app/src/main/java/com/teamcaffeine/hotswap/LoginActivity.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+        btnEmailLogin.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                Intent intent = new Intent(LoginActivity.this, EmailPasswordActivity.class);
+                startActivity(intent);
+            }
+        });
+
+//        // Set up ListView and Adapter
```
Comment (alexrosenfeld10) [inline app/src/main/java/com/teamcaffeine/hotswap/LoginActivity.java]: Remove

File: app/src/main/java/com/teamcaffeine/hotswap/HomeActivity.java (hunk 9 above)
Comment (jschadaway) [reply app/src/main/java/com/teamcaffeine/hotswap/HomeActivity.java]: Most of the data is in the Firebase mAuth instance, which is singleton across the whole app - you can access the instance from the next activity to parse login details.

**Pull request merged**